sw/host/ot_certs/src/cwt.rs

```diff
@@ -0,0 +1,911 @@
+        match &self.value {
+            CodegenVarValue::ByteArray(v) => v.is_some(),
+            CodegenVarValue::String(v) => v.is_some(),
+            CodegenVarValue::Integer(v) => v.is_some(),
+        }
+    }
+
+    fn variable_type(&self) -> VariableType {
+        match &self.value {
+            CodegenVarValue::ByteArray(_) => VariableType::ByteArray,
+            CodegenVarValue::String(_) => VariableType::String,
+            CodegenVarValue::Integer(_) => VariableType::Integer,
+        }
+    }
+
+    fn declarations(&self, prefix: &str) -> String {
+        assert!(!self.is_constant());
+
+        match self.value {
+            CodegenVarValue::ByteArray(_) => {
+                indoc::formatdoc! { r#"
+                    {prefix}uint8_t *{name};
+                    {prefix}size_t {name}_size;
+                    "#,
+                    name = self.name
+                }
+            }
+            CodegenVarValue::String(_) => {
+                indoc::formatdoc! { r#"
+                    {prefix}char *{name};
+                    {prefix}size_t {name}_size;
+                    "#,
+                    name = self.name
+                }
+            }
+            CodegenVarValue::Integer(_) => indoc::formatdoc! { r#"
+                {prefix}int64_t {name};
+                "#,
+                name = self.name
+            },
+        }
+    }
+
+    fn value_expression(&self) -> String {
+        assert!(!self.is_constant());
+        format!("values->{}", self.name)
+    }
+
+    fn size_expression(&self) -> String {
+        assert!(!self.is_constant());
+        if self.variable_type() != VariableType::Integer {
+            format!("values->{}_size", self.name)
+        } else {
+            format!("cbor_calc_arg_int_size(values->{})", self.name)
+        }
+    }
+}
+
+// CodegenStructure are stored inside a vector, and therefore use indices to refer to other items.
+// The `Var` variant represents a leaf node, while other variants represent different types of internal nodes.
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum CodegenStructure<'a> {
+    Var(&'a CodegenVar),
+    CborByteArray(usize),
+    Map(Vec<(usize, usize)>),
+    Array(Vec<usize>),
+}
+
+impl CodegenStructure<'_> {
+    fn from_template<'a>(
+        template: &CwtTemplate,
+        vars: &'a CodegenVarTable,
+    ) -> Result<Vec<CodegenStructure<'a>>> {
+        let mut nodes = Vec::<CodegenStructure>::new();
+        let mut id_mapping = HashMap::<String, usize>::new();
+
+        // This function is used to get the index of the given variable name if exists,
+        // otherwise it will create a new node inside the vector and return the index.
+        //
+        // This means that if a variable or a constant is used multiple times,
+        // they will share the same node inside the vector to avoid repetitions.
+        let mut get_or_insert = |var: Option<&str>, node: CodegenStructure<'a>| -> usize {
```

Comment:
I find the name `get_or_insert` to be somewhat misleading. In particular, the behaviour for variables is quite different from the one for variables. Also I am not sure that you use the returned index anywhere?
I think it would make more sense to call this function only for variables and drop the `node` argument. For nodes, just insert directly in `build_codegen_structure`?

Reply:
I change to directly pass those variables instead of using a closure, and then define helper closures inside the function. Hope this would make it clearer.